### PR TITLE
allow ACL policies to be associated with workload identity

### PR DESCRIPTION
**Pull request by tgross**
The original design for workload identities and ACLs allows for operators to
extend the automatic capabilities of a workload by using a specially-named
policy. This has shown to be potentially unsafe because of naming collisions, so
instead we'll allow operators to explicitly attach a policy to a workload
identity.

This changeset adds workload identity fields to ACL policy objects and threads
that all the way down to the command line. It also a new secondary index to the
ACL policy table on namespace and job so that claim resolution can efficiently
query for related policies.

---

Fixes https://github.com/hashicorp/nomad/issues/13995

cc @schmichael @apollo13 @angrycub ~~I've got this in draft while I do some end-to-end verification~~ but the overall design seems sound enough for a first look if you've got comments.

### CODE REVIEW

File: .changelog/14140.txt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+cli: `acl policy info` output format has changed to improve readability with large policy documents
```
Comment (tgross) [inline .changelog/14140.txt:2]: This is the only bit of this that's changing already-shipped code. If we want to be fussy we can pull this out to its own PR but it's a tiny change that won't get backported so seemed safe to keep here.

**Comment by tgross** [comment]
Some end-to-end testing:

```
$ WI=<token extracted from a running job=httpd>
$ NOMAD_TOKEN=$WI nomad operator api '/v1/vars'
Permission denied%

$ echo '{"Items": {"name": "Tim"}}' | nomad operator api -X PUT "/v1/var/example"
$ nomad operator api '/v1/var/example?namespace=default'
{"CreateIndex":34,"CreateTime":1660675347090671505,"Items":{"name":"Tim"},"ModifyIndex":34,"ModifyTime":1660675347090671505,"Namespace":"default","Path":"example"}%

# not allowed, as expected
$ NOMAD_TOKEN=$WI nomad operator api '/v1/var/example?namespace=default'
Permission denied%

# attach the operator policy to the job
$ nomad acl policy apply -namespace default -job httpd operator ./operator.hcl
Successfully wrote "operator" ACL policy!

$ NOMAD_TOKEN=$WI nomad operator api '/v1/var/example' | jq .
{
  "CreateIndex": 34,
  "CreateTime": 1660675347090671600,
  "Items": {
    "name": "Tim"
  },
  "ModifyIndex": 34,
  "ModifyTime": 1660675347090671600,
  "Namespace": "default",
  "Path": "example"
}

# add a job-specific secret
$ echo '{"Items": {"name": "Tim"}}' | nomad operator api -X PUT "/v1/var/nomad/jobs/httpd"

# read it as expected
$ NOMAD_TOKEN=$WI nomad operator api '/v1/var/nomad/jobs/httpd' | jq .
{
  "CreateIndex": 40,
  "CreateTime": 1660675566715335400,
  "Items": {
    "name": "Tim"
  },
  "ModifyIndex": 40,
  "ModifyTime": 1660675566715335400,
  "Namespace": "default",
  "Path": "nomad/jobs/httpd"
}
```

File: api/acl.go [hunk 2]
```diff
@@ -212,11 +212,15 @@ type ACLPolicyListStub struct {
...
-	Description string
-	Rules       string
-	CreateIndex uint64
-	ModifyIndex uint64
+	Name         string
+	Description  string
+	Rules        string
+	JobNamespace string
+	JobID        string
+	Group        string
```
Comment (apollo13) [inline api/acl.go:220]: Nitpicking but having `Namespace` and `ID` prefixed with `Job` and `Group` and `Task` not feels kinda weird.

File: nomad/structs/structs.go [hunk 3]
```diff
@@ -11841,6 +11851,19 @@ func (a *ACLPolicy) Validate() error {
...
+		mErr.Errors = append(mErr.Errors, err)
+	}
+	if a.JobID == "" && a.Group != "" {
+		err := fmt.Errorf("job ID must be set to set group")
+		mErr.Errors = append(mErr.Errors, err)
+	}
+	if a.Group == "" && a.Task != "" {
+		err := fmt.Errorf("group must be set to set task")
+		mErr.Errors = append(mErr.Errors, err)
+	}
```
Comment (shoenig) [inline nomad/structs/structs.go:11865]: same if-condition ordering nit from above

**Review by shoenig — APPROVED** [review]
LGTM! just the usual suggestions

File: api/acl.go (hunk 2 above)
Comment (tgross) [reply api/acl.go:220]: I like that, it makes it obvious that this is a special use (and has a few words fewer memory allocations for the common case).

File: command/acl_policy_apply.go [hunk 4]
```diff
@@ -89,11 +111,31 @@ func (c *ACLPolicyApplyCommand) Run(args []string) int {
...
+		return 1
+	}
+	if jobID == "" && group != "" {
+		c.Ui.Error("-job is required if -group is set")
+		return 1
+	}
+	if group == "" && task != "" {
+		c.Ui.Error("-group is required if -task is set")
+		return 1
+	}
```
[the remark this answers is not in the document]
Comment (tgross) [reply command/acl_policy_apply.go:128]: Ok, will fix.

File: api/acl.go (hunk 2 above)
Comment (tgross) [reply api/acl.go:220]: Now I'm remembering why I didn't do this originally, which is that the state store index schema was super weird. Lemme spend a little bit to see if I can figure out how to do it. Indexing on `"JobACL.Namespace"` actually does the right thing, but only if `JobACL` is non-nil so it'll break everywhere else.

Comment (tgross) [reply api/acl.go]: Just needed a custom indexer. Done.
